cadc-gms/src/main/java/org/opencadc/auth/StandardIdentityManager.java

```diff
@@ -306,75 +327,201 @@ public String toDisplayString(Subject subject) {
-
-                        InputStream istream = get.getInputStream();
-                        String str = StringUtil.readFromInputStream(istream, "UTF-8");
-                        JSONObject json = new JSONObject(str);
-                        String sub = json.getString("sub");
-                        String username = json.getString("preferred_username");
-                        // TODO: register an X509 DN with IAM and see if I can get it back here
-
-                        OpenIdPrincipal oip = new OpenIdPrincipal(oidcClient.getIssuerURL(), sub);
-                        HttpPrincipal hp = new HttpPrincipal(username);
-
-                        s.getPrincipals().remove(raw);
-                        s.getPrincipals().add(oip);
-                        s.getPrincipals().add(hp);
-
-                        AuthorizationToken authToken = new AuthorizationToken(challengeType, credentials, oidcDomains, oidcScope);
-                        s.getPublicCredentials().add(authToken);
-                    } catch (NotAuthenticatedException ex) {
-                        JSONObject json = new JSONObject(ex.getMessage());
-                        String error = json.getString("error");
-                        String details = json.getString("error_description");
-                        // details usually includes the invalid access token: truncate
-                        StringBuilder sb = new StringBuilder(error);
-                        sb.append(" reason: ");
-                        int max = Math.min(details.length(), 32);
-                        sb.append(details.subSequence(0, max));
-                        if (max < details.length()) {
-                            sb.append("...");
-                        }
-                        throw new NotAuthenticatedException(challengeType, NotAuthenticatedException.AuthError.INVALID_TOKEN, sb.toString());
-                    } catch (Exception ex) {
-                        throw new NotAuthenticatedException(challengeType, NotAuthenticatedException.AuthError.INVALID_TOKEN, ex.getMessage(), ex);
-                    }
+    private void validateToken(Subject s, AuthorizationTokenPrincipal raw) {
+        // this method validates a single token using either the issuer public key (preferred because the key is
+        // cached locally) or by calling the user info endpoint and updates the subject accordingly
+        String[] tval = raw.getHeaderValue().split(" ");
+        String challengeType = tval[0];
+        String credentials = tval[1];
+        log.debug("challenge type: " + challengeType);
+        log.debug("credentials: " + credentials);
+
+        URI jwtIssuer = null;
+        try {
+            jwtIssuer = getJwtIssuer(credentials);
+            if (!jwtIssuer.normalize().equals(oidcClient.issuer.normalize())) {
+                log.debug("Token from untrusted issuer: " + jwtIssuer + " ignored");
+                jwtIssuer = null; // not the configured issuer, so don't use it
```

Comment:
by setting this to null, the validation falls through to trying userinfo endpoint on the trusted IDP... is that necessary if the this is a JWT and the issuer doesn't match? Should this throw NotAuthenticatedException right here?